mobius_model.h

```diff
@@ -1,1502 +0,0 @@
-
-#if !defined(MOBIUS_MODEL_H)
-
-
-//NOTE: The purpose of having unit_h, input_h, equation_h etc. be structs that contain a numeric handle rather than just letting them be a handle directly is that we can then use the C++ type system to get type safety. Unfortunately, C++ does not allow you to typedef a unique copy of a type that is not interchangable with others. However the type system WILL distinguish between two differently named structs even though they are otherwise equal.
-
-typedef u32 entity_handle;
-
-#define MODEL_ENTITY_HANDLE(Type) struct Type \
-{ \
-	entity_handle Handle; \
-}; \
-bool operator==(const Type &A, const Type &B) { return A.Handle == B.Handle; } \
-bool operator!=(const Type &A, const Type &B) { return A.Handle != B.Handle; } \
-bool operator<(const Type &A, const Type &B) { return A.Handle < B.Handle; } \
-inline bool IsValid(Type H) { return H.Handle > 0; }
-
-MODEL_ENTITY_HANDLE(unit_h)
-
-MODEL_ENTITY_HANDLE(input_h)
-MODEL_ENTITY_HANDLE(equation_h)
-MODEL_ENTITY_HANDLE(parameter_double_h)
-MODEL_ENTITY_HANDLE(parameter_uint_h)
-MODEL_ENTITY_HANDLE(parameter_bool_h)
-MODEL_ENTITY_HANDLE(parameter_time_h)
-
-MODEL_ENTITY_HANDLE(solver_h)
-
-MODEL_ENTITY_HANDLE(index_set_h)
-
-MODEL_ENTITY_HANDLE(parameter_group_h)
-
-#undef MODEL_ENTITY_HANDLE
-
```

Comment:
I suggest to use templates here, see here:https://softwareengineering.stackexchange.com/questions/243154/c-strongly-typed-typedef

Reply:
What would that solve compared to the current way though? I am just afraid that it would skyrocket the compilation times. Could try it of course.

Reply:
It follows best practices: https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros
and it is more readable. I doupt compilation overhead will be at a detectable level.

Reply:
Again, I disagree with there being any "best practices" defined by an arbitrary site. If you claim to follow C++ "best practices", you would probably not even get a job in certain companies. There are a lot of different experiences tied to different use cases. I can make a change if there is a good reason for it that can be explained and that is relevant for this particular project. I use C++ templates as little as possible because they usually cause me more problems than macros. In this case it is even a very self-contained macro that is undefed right after it is being used, so it does not leak out. I also think this particular example is pretty trivial, so it would probably not matter too much either way.